### ISSUE TITLE
wreck: cpu affinity support requires OS ids for cores

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#1591]
From a conversation this AM, an idea to run 1 flux-broker per socket was proposed as a short term fix for properly binding tasks to their local GPUs.

However, this may not work as expected because current wreck binding assumes the `core` list in `R_lite` correspond to OS core index, not a logical index.

If this support is broken, we need to determine the most time efficient fix.

**Comment by dongahn** [comment]
Thanks. I just sent an email to Langer about this.

> From a conversation this AM, an idea to run 1 flux-broker per socket was proposed as a short term fix for properly binding tasks to their local GPUs.

I believe this goes beyond this use case.

Because our scheduler generates cpu list from hwloc's logical index, not from os index, and wreck's `set-affinity` expects os index, our set affinity is just broken at the moment. 

I can see we fix this within sched or within wreck.

**Comment by grondo** [comment]
oops, I had forgotten how CPU affinity was implemented in wrexecd. We're actually using `hwloc_set_cpubind()` *not* `sched_setaffinity(2)` (that was the older implementation).

Here's the current code used in `wrexecd` to bind all tasks to the core list in `R_lite`:

```C
static int do_hwloc_core_affinity (struct prog_ctx *ctx)
{
    int rc = -1;
    int depth, i;
    hwloc_topology_t topology;
    hwloc_cpuset_t coreset = NULL;
    hwloc_cpuset_t resultset = NULL;
    flux_t *h = ctx->flux;

    if (hwloc_topology_init (&topology) < 0) {
        flux_log_error (h, "hwloc_topology_init");
        return (-1);
    }
    if (hwloc_topology_load (topology) < 0) {
        flux_log_error (h, "hwloc_topology_load");
        goto out;
    }
    if (!(coreset = hwloc_bitmap_alloc ())
       || !(resultset = hwloc_bitmap_alloc ())) {
        flux_log_error (h, "hwloc_bitmap_alloc");
        goto out;
    }
    if (hwloc_bitmap_list_sscanf (coreset, ctx->rankinfo.cores) < 0) {
        flux_log_error (h, "hwloc_sscanf(%s)", ctx->rankinfo.cores);
        goto out;
    }
    depth = hwloc_get_type_depth (topology, HWLOC_OBJ_CORE);
    if (depth == HWLOC_TYPE_DEPTH_UNKNOWN
       || depth == HWLOC_TYPE_DEPTH_MULTIPLE) {
        flux_log_error (h, "hwloc_get_type_depth (CORE)");
        goto out;
    }
    i = hwloc_bitmap_first (coreset);
    while (i >= 0) {
        hwloc_obj_t core = hwloc_get_obj_by_depth (topology, depth, i);
        if (core)
            hwloc_bitmap_or (resultset, resultset, core->cpuset);
        else
            flux_log_error (h, "hwloc_get_obj_by_depth: core%d", i);
        i = hwloc_bitmap_next (coreset, i);
    }
    if ((rc = hwloc_set_cpubind (topology, resultset, 0)) < 0)
        flux_log_error (h, "hwloc_set_cpubind: %s", strerror (errno));
out:
    hwloc_bitmap_free (resultset);
    hwloc_bitmap_free (coreset);
    hwloc_topology_destroy (topology);
    return (0);
}
```

I would have to pore over the docs to determine if this code is doing the right thing or not.

In any event, if I implemented a plugin to do the `per-task` binding, it would use `sched_setaffinity` to divide up the current cpu affinity to each task.

Sorry for confusing this issue. I'm willing to change the code above to whatever will work in the short term.

**Comment by grondo** [comment]
@SteVwonder, I wanted to make sure you saw the above. If I understand the documentation correctly `hwloc_get_obj_by_depth` returns the object at *logical index* `i`, so perhaps we don't need to switch sched to OS index at this point? I think we'd need to do some experiments with libhwloc constrained to various cpusets and ensure the above code still does what we expect. And there is also the question of the numbering of devices in CUDA_VISIBLE_DEVICES -- we need to ensure the ids in the `R_lite` gpu list correspond exactly to what should appear in CUDA_VISIBLE_DEVICES.

**Comment by dongahn** [comment]
@grondo:

> hwloc_set_cpubind ()

So this function works on the logical id? If so, then we should be good, and Langer's case is already handled (Though some tests seem warranted)... Thoughts?

> In any event, if I implemented a plugin to do the per-task binding, it would use sched_setaffinity to divide up the current cpu affinity to each task.

Hmmm. Why can't `hwloc_set_cpubind ()` be used for per-task binding as well? If we can, this may save us having to resolve logical to os index -- just relying on hwloc to do this translation for us.

**Comment by SteVwonder** [comment]
Thanks @grondo.  The results in https://github.com/flux-framework/flux-sched/issues/367 make a lot more sense now.  I'll try a few more cases besides the one's I already tried.  Specifically, Langer's use-case, as mentioned by @dongahn.

> we need to ensure the ids in the R_lite gpu list correspond exactly to what should appear in CUDA_VISIBLE_DEVICES.

I'll look more into the numbering of the devices in HWLOC and how that maps to CUDA_VISIBLE_DEVICES.  I know there is another environment variable that can be set to change CUDA's interpretation of the IDs, and hopefully one of the interpretations maps to HWLOC's numbering.

**Comment by grondo** [comment]
> Hmmm. Why can't hwloc_set_cpubind () be used for per-task binding as well? If we can, this may save us having to resolve logical to os index -- just relying on hwloc to do this translation for us.

Yes, I would probably use libhwloc to do per-task binding. I specifically meant if we used a *plugin* to do it, we'd have to use sched_setaffinity, since that is all that is currently available from the Lua side.

I can work on an implementation in wrexecd.c instead though.

**Comment by dongahn** [comment]
> Yes, I would probably use libhwloc to do per-task binding. I specifically meant if we used a plugin to do it, we'd have to use sched_setaffinity, since that is all that is currently available from the Lua side.

Ah... Yeah there is no hwloc lua binding...

**Comment by dongahn** [comment]
Here is the xml from hwloc from Sierra. At least it looks like cuda coprocessor device is hung off of NUMANODE (socket), the topological distance will be preserved when the scheduler do depth first walk on the hwloc tree.

https://github.com/flux-framework/flux-sched/blob/master/t/data/hwloc-data/004N/exclusive/04-brokers-sierra2/0.xml#L339

https://github.com/flux-framework/flux-sched/blob/master/t/data/hwloc-data/004N/exclusive/04-brokers-sierra2/0.xml#L349

Still, @grondo's question needs to be confirmed: whether the cuda device number used there correspond to what `CUDA_VISIBLE_DEVICE` expects.

**Comment by dongahn** [comment]
From [this posting](https://www.mail-archive.com/hwloc-users@lists.open-mpi.org/msg00986.html), it looks like hwloc's cuda device numbering matches with nvidia-smi order, which is PCI order. But from [this page](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars), `CUDA_VISIBLE_DEVICES` by defaults uses `CUDA_​DEVICE_​ORDER=FASTEST_FIRST`. 

So I have to believe, if our gpu binding also sets `CUDA_​DEVICE_​ORDER=PCI_BUS_ID`, we should be good across the board.

**Comment by grondo** [comment]
Thanks @dongahn! Can you open an issue and tie to 0.10.0?
